### ISSUE TITLE
[HOLD for payment 2024-06-06] [$250] [Search v1] Search page scroll is not smooth

### ISSUE THREAD

**Issue by luacmartins** [Expensify/App#41581]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:**
**Slack conversation:**

## Action Performed:
1. Open the search page, `Profile > Troubleshoot > New Search Page`
2. Scroll through the list of results
3. Notice that the items flicker and on mobile the list jumps around

## Expected Result:
Scroll should be smooth

## Actual Result:
Scroll is not smooth

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/22219519/d0a06780-395b-4bfb-9dbd-6242be6db2d8

https://github.com/Expensify/App/assets/22219519/89b5fcfd-293b-4373-8687-add4cf8ed808

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01fb9b7cf710602d40</li>
        <li>Upwork Job ID: 1786403109998579712</li>
        <li>Last Price Increase: 2024-05-03</li>
<li>Automatic offers: </li>
<ul>
<li>eh2077 | Reviewer | 0</li>
<li>charles-liang | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @muttmuure</details>

**Comment by nharsh94** [comment]
Hi @expensify team,

[Here's](https://codepen.io/Noah-Harshbarger/pen/poBXxda) a link to a solution to the code using JS/TS. It should be the label "smooth scroll test" with three different colors for reference. Please let me know what you think. 

I tried to keep it in TS for **customization**, **ease of integration** (kept it to a class encapsulation), **no dependencies**, and **performance** for cross-browser compatibility and maintainability. I only used DOM APIs and JS features, so it should be pretty durable. No external libraries or frameworks, as well.

**Comment by eh2077** [comment]
@nharsh94 Thanks for your proposal. I'm not following your root cause analysis. Can you elaborate on why the scroll is not smooth?

**Comment by nharsh94** [comment]
Hi @eh2077, 

Apologies -- I went through the `Search` page, and I think it may come down to heavy rendering, expensive operations, and excessive re-renders for the search function. As the items are popping up, it's bottlenecking the rendering pipeline. I have a few more options that I found since reading through again:

1. With React already implemented, [memoization](https://react.dev/reference/react/memo) could be used to only change props that have been rendered when different (instead of re-rendering when the page refreshes to grab the other search items). This could keep the remaining list on top the same while the rest of the items are loading. 

2. Since the lists for searching can be a lot larger, the implementation of [react-virtualized](https://www.npmjs.com/package/react-virtualized-auto-sizer) would only render what's visible on the viewport, reducing the number of DOM elements and improving the scroll performance.

I thought that it was only due to the amount of search items available (and could be fixed with a Class), but with the mobile video it seems more like a performance bottlenecking issue.

**Comment by eh2077** [comment]
@nharsh94 Thanks for the update, but your root cause analysis still seems vague to me.

**Comment by eh2077** [comment]
Still looking for better proposals

**Comment by luacmartins** [comment]
Still looking for proposals

**Comment by eh2077** [comment]
@charles-liang Thanks for your proposal. It's good to see you tried to pinpoint the root cause by profiling the page. 

> 2. By removing all elements of the item and replacing them with a single text element, the delay issue was improved.

Do you mean the `TransactionListItem` here?

> 5. All functions returning JSX were removed from TransactionListItem.
> 6. Analyzed again using React Dev Tools, the previously observed delay has disappeared.

Can you elaborate on why those functions cause the delay when scrolling?

**Comment by eh2077** [comment]
@charles-liang Thanks for the update. Can you detail your improvement plan in the third/solution part of your proposal?

**Comment by eh2077** [comment]
@charles-liang I mean the main solution

> removed All functions returning JSX from TransactionListItem.

This part can be more detail.

**Comment by eh2077** [comment]
Can you add them to your proposal? Also it'll be very helpful if you can provide a branch to test.

**Comment by eh2077** [comment]
@charles-liang Can you add a summary about the root cause of the issue and update your proposal? And please elaborate on what changes are needed in the solution part. Note please explain them in plain english instead of a test branch.

**Comment by eh2077** [comment]
> @charles-liang Thanks for the udpate.
> 
> > The root cause is that every time you scroll, it removes the TransactionListItem that leaves the visible area and renders new TransactionListItems.
> 
> This seems like the key factor causing the scrolling issue. Can you elaborate on why `TransactionListItem`s are not cached?

You didn't answer the question - Can you explain why the _**list of `TransactionListItem`s**_ are not cached when scrolling?

**Comment by eh2077** [comment]
@charles-liang Can you update your proposal based on the new findings above?

**Comment by eh2077** [comment]
@charles-liang Thanks for the update.

> Setting windowSize and updateCellsBatchingPeriod to the appropriate value in SelectionList in Search.tsx.

Can we get significant improvement in both the profiling results and the smoothness of scrolling after tuning these parameters?

**Comment by eh2077** [comment]
@charles-liang Thanks for your patience during the discussion.

---
I think the root cause of this unsmooth scrolling issue is uncovered
- The current settings of [BaseSelectionList.tsx](https://github.com/Expensify/App/blob/807c945ef8edabe86cc62f16c866aa754db8750b/src/components/SelectionList/BaseSelectionList.tsx#L598) can't efficiently render `TransactionListItem`s when scrolling, see [comment](https://github.com/Expensify/App/issues/41581#issuecomment-2100884045) and [comment](https://github.com/Expensify/App/issues/41581#issuecomment-2101841087)
-  It's relatively expensive to render `TransactionListItem`s

---

It's great that @charles-liang followed our best practice [React Performance Tips](https://github.com/Expensify/App/blob/main/contributingGuides/PERFORMANCE.md) to profile the performance issue and adopt tips, like using [PureComponent](https://reactjs.org/docs/react-api.html#reactpurecomponent) and [React.memo()](https://reactjs.org/docs/react-api.html#reactmemo), to optimise it, though they seem new to this project and this is likely to be their first contribution.
We can also refer to [Optimizing Flatlist Configuration](https://reactnative.dev/docs/optimizing-flatlist-configuration)

---

That said, I think we can go with @charles-liang 's [proposal](https://github.com/Expensify/App/issues/41581#issuecomment-2097784758). Coding details can be discussed in PR.

🎀👀🎀 C+ reviewed

**Comment by luacmartins** [comment]
This issue is yours @charles-liang. Let me know when we can expect a PR.

**Comment by eh2077** [comment]
@charles-liang Kindly comment an eta for the PR, so we can keep the team updated on the progress. Thanks

**Comment by eh2077** [comment]
@charles-liang Thank you. Feel free to let us know if you have any questions.

**Comment by eh2077** [comment]
Still working on the PR

**Comment by eh2077** [comment]
### Checklist

- [x] [@eh2077] The PR that introduced the bug has been identified. Link to the PR: _**This is a performance issue not a bug, so there isn't a specific PR responsible for this issue**_
- [x] [@eh2077] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: N/A
- [x] [@eh2077] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: _**N/A, we already have our best practices documented here [React Performance Tips](https://github.com/Expensify/App/blob/main/contributingGuides/PERFORMANCE.md)**_
- [x] [@eh2077] Determine if we should create a regression test for this bug. _**No**_

**Comment by luacmartins** [comment]
Are we just missing payment here?

**Comment by eh2077** [comment]
@luacmartins Yes, that's right. Just asking, is it possible to increase the bounty a little bit as it needed more effort than usual bug fixes. I'm fine if we just do with the default bounty. Thanks

**Comment by luacmartins** [comment]
Thanks for confirming! I think the $250 bounty is still fair in this case.

**Comment by eh2077** [comment]
No problem! Let's just wait for BZ team to issue the payment.

cc @muttmuure

**Comment by muttmuure** [comment]
Looking

**Comment by muttmuure** [comment]
Paid
